Add spec for obsoletions in .NET 5 (#139)
dotnet/designs: accepted/2020/better-obsoletion/obsoletions-in-net5.md

the `ObsoleteAttribute` gained new `DiagnosticId` and `UrlFormat` properties. With
these properties, obsoletions can be grouped together when multiple APIs are
marked as obsolete for the same reason. This capability is reflected in the list
of obsoletions below. All of these obsoletions will use the `UrlFormat` of
`https://aka.ms/dotnet-warnings/{0}`.

1. The Constrained Execution Region (CER) feature is no longer implemented
    * DiagnosticId: `MSLIB0004`
    * APIs:
        * `System.Runtime.ConstrainedExecution.PrePrepareMethodAttribute`
        * `System.Runtime.ConstrainedExecution.ReliabilityContractAttribute`
        * `System.Runtime.ConstrainedExecution.Cer`
        * `System.Runtime.ConstrainedExecution.Consistency`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegions`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegionsNoOP`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ProbeForSufficientStack`
1. This AppDomain method is no longer implemented
    * DiagnosticId: `MSLIB0005`
    * APIs:

>         * `PrepareContractedDelegate`

Reviewer: This should be in the __CER__ bucket above.

With `dotnet/runtime` PR [#33248](https://github.com/dotnet/runtime/pull/33248),
the `ObsoleteAttribute` gained new `DiagnosticId` and `UrlFormat` properties. With
these properties, obsoletions can be grouped together when multiple APIs are
marked as obsolete for the same reason. This capability is reflected in the list
of obsoletions below. All of these obsoletions will use the `UrlFormat` of
`https://aka.ms/dotnet-warnings/{0}`.

1. The Constrained Execution Region (CER) feature is no longer implemented
    * DiagnosticId: `MSLIB0004`
    * APIs:
        * `System.Runtime.ConstrainedExecution.PrePrepareMethodAttribute`
        * `System.Runtime.ConstrainedExecution.ReliabilityContractAttribute`
        * `System.Runtime.ConstrainedExecution.Cer`
        * `System.Runtime.ConstrainedExecution.Consistency`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegions`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegionsNoOP`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ProbeForSufficientStack`

> 2. This App Domain method is no longer implemented

Reviewer 2: ```suggestion
2. This AppDomain method is no longer implemented
```

    * DiagnosticId: `MSLIB0005`
    * APIs:
        * `PrepareContractedDelegate`
2. The Global Assembly Cache is no longer implemented
    * DiagnosticId: `MSLIB0006`
    * APIs:
        * `System.Reflection.Assembly.GlobalAssemblyCache`

> 3. `System.Net.WebSockets.WebSocket.IsApplicationTargeting45` is no longer applicable

Reviewer 2: This is already marked Obsolete.  Are we saying we're just going to give it a diagnostic id / message?

Author: Ah, I missed that. It's probably not worth it to add the id/url for this one (as we'd then need to have a page for the URL too).  I'll remove this one.

    * DiagnosticId: `MSLIB0007`
    * APIs:
        * `System.Net.WebSockets.WebSocket.IsApplicationTargeting45`
4. `Thread.Abort` is no longer supported and throws PlatformNotSupportedException
    * DiagnosticId: `MSLIB0008`
    * APIs:
        * `System.Threading.Thread.Abort()`
        * `System.Threading.Thread.Abort(Object)`
5. The default implementations of these cryptography algorithms are no longer supported and throw PlatformNotSupportedException
    * DiagnosticId: `MSLIB0009`
    * APIs:
        * `System.Security.Cryptography.SymmetricAlgorithm.Create()`
        * `System.Security.Cryptography.AssymetricAlgorithm.Create()`
        * `System.Security.Cryptography.HMAC.Create()`
        * `System.Security.Cryptography.KeyedHashAlgorithm.Create()`

> 6. The CreatePbdGenerator API is no longer supported and throws PlatformNotSupportedException

Reviewer 2: I don't see this method anywhere in dotnet/runtime.

Author: There was a typo (it should be Create**Pdb**Generator). Sorry about that.

* The ref is in `System.Linq.Expressions\ref\System.Linq.Expressions.cs` in the `DebugInfoGenerator` class
* The implementation is in `System.Linq.Expressions\src\System\Runtime\CompilerServices\DebugInfoGenerator.cs`

    * DiagnosticId: `MSLIB0010`
    * APIs:
        * `System.Runtime.CompilerServices.DebugInfoGenerator.CreatePbdGenerator`
7. The AuthenticationManager Authenticate and PreAuthenticate APIs are no longer supported and throw PlatformNotSupportedException
    * DiagnosticId: `MSLIB0011`
    * APIs:
        * `System.Net.AuthenticationManager.Authenticate`
        * `System.Net.AuthenticationManager.PreAuthenticate`
8. These Remoting APIs are no longer supported and throw PlatformNotSupportedException
    * DiagnosticId: `MSLIB0012`
    * APIs:
        * `System.MarshalByRefObject.GetLifetimeService`
        * `System.MarshalByRefObject.InitializeLifetimeService`
9. Code Access Security (CAS) APIs are no longer supported
    * DiagnosticId: `MSLIB0013`
    * APIs: (all from the `System.Security.Permissions` namespace)

>         * Classes: (everything *except* `PrincipalPermission` (still implemented) and `PrincipalPermissionAttribute` (already obsolete))

Reviewer 3: Is the fact that it happens to be implemented as compat quirk a good reason for not obsoleting it? I think it would be better to obsolete it too.

Author: @GrabYourPitchforks - what do you think?

Reviewer: I'm fine with obsoleting it (as warning). Just didn't want to push the issue since it wasn't a "slam dunk" like the other APIs, which all either no-op or throw PNSE.

Author: @GrabYourPitchforks Two questions:

1. Should we group all of the CAS stuff into the same DiagnosticId already in use for `PrincipalPermissionAttribute` instead of creating a new one?
1. Should we use a different DiagnosticId for `PrincipalPermission` since it's less of a slam dunk?

Reviewer: For (1), yes. Honestly there wasn't much of a point to creating a standalone diagnostic code just for `PrincipalPermissionAttribute` since it's marked _as error_ and you can't suppress it. So may as well reuse that same code for this bucket.

For (2), we plan on having a fixer at some point. IMO we should only use a different code if doing so would make the fixer easier to write / apply. I'm not really familiar with that feature, though, so somebody like @bartonjs or @stephentoub might know more.

Author: Thanks. For (2), that is enough for me to want to use a separate DiagnosticId for `PrincipalPermission`. It can be suppressed separately then too.

Author: Leaving this unresolved to give @bartonjs and @stephentoub a chance to weigh in. But I've now assigned all of the CAS items into the existing `MSLIB0002` DiagnosticId and then used the new `MSLIB0012` for `PrincipalPermission` and `PermissionState`.

Reviewer 4: Fixers are activated by diagnostic IDs, merely having the right ID makes the lightbulb show up (I'm given to understand).  So anything with a fixer gets a different code than things without a fixer.  (If one fixer can fix more than one sub-problem but have a sensible display name, one code is fine.  Or it can activate over multiple codes.)

Reviewer 4: This also effectively means anything bucketed into MSLIB0012 now can never gain a fixer.

Reviewer 3: > This also effectively means anything bucketed into MSLIB0012 now can never gain a fixer.

I am not following. Can you expand on this?

Author: @mavasani Could you help us make sure we understand this fully? If multiple conditions share a DiagnosticId, and then we later want to offer a fixer for _one_ of those conditions, would we be able to?  As @bartonjs and I understand it, DiagnosticId values need to be unique for diagnostics with fixers available.

I'm also thinking about suppression for this too.  Since `PrincipalPermission` might have use cases remaining when the other CAS-related APIs don't, I wouldn't want someone to suppress `PrincipalPermission` and inadvertently suppress all other API obsoletions from the CAS category.

Reviewer 5: > If multiple conditions share a DiagnosticId, and then we later want to offer a fixer for one of those conditions, would we be able to

Diagnostics can attach a custom set of key-value pair `properties` to itself, and fixer can use these properties to determine whether or not to fix a specific diagnostic. This enables reporting multiple diagnostics with same ID, when fix is desired to be offered only for a subset of these.

https://github.com/dotnet/roslyn/blob/7023b0bbd4ca161ff6005b1219da1bc60b0fa58e/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L98-L102

Author: Thanks, @mavasani. The `ObsoleteAttribute` diagnostics probably can't do this though, since Roslyn is generating the diagnostics from the `ObsoleteAttribute` for us. We'd have no way of attaching the key-value pairs (that I can think of).

Reviewer 5: > I'm also thinking about suppression for this too. Since PrincipalPermission might have use cases remaining when the other CAS-related APIs don't, I wouldn't want someone to suppress PrincipalPermission and inadvertently suppress all other API obsoletions from the CAS category.

Suppression/Disabling an ID is a valid scenario of concern - currently, Roslyn compiler only supports disabling diagnostic set by IDs, a finer granularity configuration where a specific sub-bucket of diagnostics with shared ID can be turned off is not yet supported. @sharwell has proposed this feature here: https://github.com/dotnet/roslyn/issues/27712, but this needs design approval from Roslyn compiler team first.

Reviewer 3: > `PrincipalPermission` might have use cases remaining 

Those use look pretty weak. I would not feel bad about people having to add warning disable around those if they want to keep them.

# Obsoletions in .NET 5

**PM** [Immo Landwerth](https://github.com/terrajobst)

For .NET 5, [we're making](better-obsoletion.md) it more viable to obsolete APIs.
In this document, we're describing which APIs we intend to obsolete in .NET 5 and
which diagnostic IDs we will use for those.

With `dotnet/runtime` PR [#33248](https://github.com/dotnet/runtime/pull/33248),
the `ObsoleteAttribute` gained new `DiagnosticId` and `UrlFormat` properties. With
these properties, obsoletions can be grouped together when multiple APIs are
marked as obsolete for the same reason. This capability is reflected in the list
of obsoletions below. All of these obsoletions will use the `UrlFormat` of
`https://aka.ms/dotnet-warnings/{0}`.

1. The Constrained Execution Region (CER) feature is no longer implemented

>     * DiagnosticId: `MSLIB0004`

Reviewer 6: Per offline conversation I would much prefer not using any MS/Microsoft prefix for efforts like this relating to the base libraries, which are a dotnet foundation project.

Author: We're getting consensus around `SYSLIB`. I'll move forward with that.

>     * DiagnosticId: `MSLIB0004`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0004`
```

    * APIs:
        * `System.Runtime.ConstrainedExecution.PrePrepareMethodAttribute`
        * `System.Runtime.ConstrainedExecution.ReliabilityContractAttribute`
        * `System.Runtime.ConstrainedExecution.Cer`
        * `System.Runtime.ConstrainedExecution.Consistency`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegions`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegionsNoOP`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareContractedDelegate`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ProbeForSufficientStack`
1. The Global Assembly Cache is no longer implemented

>     * DiagnosticId: `MSLIB0005`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0005`
```

    * APIs:
        * `System.Reflection.Assembly.GlobalAssemblyCache`
1. `Thread.Abort` is no longer supported and throws PlatformNotSupportedException

>     * DiagnosticId: `MSLIB0006`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0006`
```

    * APIs:
        * `System.Threading.Thread.Abort()`
        * `System.Threading.Thread.Abort(Object)`
1. The default implementations of these cryptography algorithms are no longer supported and throw PlatformNotSupportedException

>     * DiagnosticId: `MSLIB0007`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0007`
```

    * APIs:
        * `System.Security.Cryptography.SymmetricAlgorithm.Create()`
        * `System.Security.Cryptography.AssymetricAlgorithm.Create()`
        * `System.Security.Cryptography.HMAC.Create()`
        * `System.Security.Cryptography.KeyedHashAlgorithm.Create()`
1. The CreatePdbGenerator API is no longer supported and throws PlatformNotSupportedException

>     * DiagnosticId: `MSLIB0008`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0008`
```

    * APIs:
        * `System.Runtime.CompilerServices.DebugInfoGenerator.CreatePdbGenerator`
1. The AuthenticationManager Authenticate and PreAuthenticate APIs are no longer supported and throw PlatformNotSupportedException

>     * DiagnosticId: `MSLIB0009`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0009`
```

    * APIs:
        * `System.Net.AuthenticationManager.Authenticate`
        * `System.Net.AuthenticationManager.PreAuthenticate`
1. These Remoting APIs are no longer supported and throw PlatformNotSupportedException

>     * DiagnosticId: `MSLIB0010`

Author: ```suggestion
    * DiagnosticId: `SYSLIB0010`
```

    * APIs:
        * `System.MarshalByRefObject.GetLifetimeService`
        * `System.MarshalByRefObject.InitializeLifetimeService`
1. Code Access Security (CAS) APIs are no longer supported

>     * DiagnosticId: `MSLIB0002` (reusing the existing DiagnosticId from `PrincipalPermissionAttribute`)

Author: ```suggestion
    * DiagnosticId: `SYSLIB0002` (reusing the existing DiagnosticId from `PrincipalPermissionAttribute`)
```

        * `System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegions`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareConstrainedRegionsNoOP`
        * `System.Runtime.CompilerServices.RuntimeHelpers.PrepareContractedDelegate`
        * `System.Runtime.CompilerServices.RuntimeHelpers.ProbeForSufficientStack`
1. The Global Assembly Cache is no longer implemented
    * DiagnosticId: `SYSLIB0005`
    * APIs:
        * `System.Reflection.Assembly.GlobalAssemblyCache`
1. `Thread.Abort` is no longer supported and throws PlatformNotSupportedException
    * DiagnosticId: `SYSLIB0006`
    * APIs:
        * `System.Threading.Thread.Abort()`
        * `System.Threading.Thread.Abort(Object)`
1. The default implementations of these cryptography algorithms are no longer supported and throw PlatformNotSupportedException
    * DiagnosticId: `SYSLIB0007`
    * APIs:
        * `System.Security.Cryptography.SymmetricAlgorithm.Create()`
        * `System.Security.Cryptography.AssymetricAlgorithm.Create()`
        * `System.Security.Cryptography.HMAC.Create()`

>         * `System.Security.Cryptography.KeyedHashAlgorithm.Create()`

Author: @bartonjs `HashAlgorithm.Create()` also throws (and there's a unit test including it in the same list as these others). Should we obsolete it as well?

Reviewer 4: > HashAlgorithm.Create() also throws (and there's a unit test including it in the same list as these others). Should we obsolete it as well?

Yep.

            * `System.Transactions.DistributedTransactionPermission`
            * `System.Web.AspNetHostingPermission`
            * `System.Xaml.Permissions.XamlLoadPermission`
        * Enums: (everything from the `System.Security.Permissions` namespace *except* `PermissionState`, which is used by `PrincipalPermission`; see below)
            * `DataProtectionPermissionFlags`
            * `EnvironmentPermissionAccess`
            * `FileDialogPermissionAccess`
            * `FileIOPermissionAccess`
            * `HostProtectionResource`
            * `IsolatedStorageContainment`
            * `KeyContainerPermissionFlags`
            * `MediaPermissionAudio`
            * `MediaPermissionImage`
            * `MediaPermissionVideo`
            * `ReflectionPermissionFlag`
            * `RegistryPermissionAccess`
            * `SecurityAction`
            * `SecurityPermissionFlag`
            * `StorePermissionFlags`
            * `TypeDescriptorPermissionFlags`
            * `UIPermissionClipboard`
            * `UIPermissionWindow`
            * `WebBrowserPermissionLevel`
1. PrincipalPermission is no longer supported
    * DiagnosticId: `MSLIB0012`
    * APIs:

>         * `System.Security.Permissions.PrincipalPermission`

Reviewer 3: @bartonjs Could you please comment on https://github.com/dotnet/designs/pull/139#discussion_r453846088 ?

I do not see why `PrincipalPermission` should be its own bucket separate from all other permissions.

Author: I had combined it in but then ended up having to keep the `PrincipalPermissionAttribute` constructor as its own ID since it's an **error** level and the others are **warnings**. But `PrincipalPermission` falls into the general CAS category.
